libecole/src/instance/capacitated-facility-location.cpp

```diff
@@ -0,0 +1,254 @@
+ * Constraints are relased automatically (through unique_ptr in scip::create_cons_basic_linear).
+ */
+auto add_demand_cons(SCIP* scip, xt::xtensor<SCIP_VAR*, 2> const& serving_vars) -> void {
+	auto const inf = SCIPinfinity(scip);
+	auto const [n_customers, n_facilities] = serving_vars.shape();
+	// Asserting row major as we pass the pointer as an array to SCIP when creating constraints
+	assert(serving_vars.layout() == xt::layout_type::row_major);
+
+	// Note change to the negative of the constraint from
+	// Gasse et al. Exact combinatorial optimization with graph convolutional neural networks 2019.
+	auto const coefs = xvector({n_facilities}, 1.);
+	for (std::size_t customer_idx = 0; customer_idx < n_customers; ++customer_idx) {
+		auto const name = fmt::format("d_{}", customer_idx);
+		auto cons = scip::create_cons_basic_linear(
+			scip, name.c_str(), n_facilities, &serving_vars(customer_idx, 0), coefs.data(), 1.0, inf);
+		scip::call(SCIPaddCons, scip, cons.get());
+	}
+}
+
+/** Add n_facilities constraints stating that facilities cannot exceed their capacity.
+ *
+ * For each facility the sum of all fraction of demand served, multiplied by the demand, must be smaller than the
+ * facility capacity.
+ * Constraints are relased automatically (through unique_ptr in scip::create_cons_basic_linear).
+ */
+auto add_capacity_cons(
+	SCIP* scip,
+	xt::xtensor<SCIP_VAR*, 2> serving_vars,
+	xt::xtensor<SCIP_VAR*, 1> const& facility_vars,
+	xvector const& demands,
+	xvector const& capacities) -> void {
+	auto const inf = SCIPinfinity(scip);
+	// Transposing and asserting row major as we pass the pointer as an array to SCIP when creating constraints.
+	serving_vars = xt::transpose(std::move(serving_vars));
+	assert(serving_vars.layout() == xt::layout_type::row_major);
+
+	auto const [n_facilities, n_customers] = serving_vars.shape();
+	assert(facility_vars.size() == n_facilities);
+	assert(demands.size() == n_customers);
+	assert(capacities.size() == n_facilities);
+
+	for (std::size_t facility_idx = 0; facility_idx < n_facilities; ++facility_idx) {
+		auto const name = fmt::format("c_{}", facility_idx);
+		auto cons = scip::create_cons_basic_linear(
+			scip, name.c_str(), n_customers, &serving_vars(facility_idx, 0), demands.data(), -inf, 0.);
+		scip::call(SCIPaddCoefLinear, scip, cons.get(), facility_vars[facility_idx], -capacities[facility_idx]);
+		scip::call(SCIPaddCons, scip, cons.get());
+	}
+}
+
+/** Add n_customers * n_facilities constraint that tighten the LP relaxation.
+ *
+ * A facility cannot serve demand if it is not open.
+ * Constraints are relased automatically (through unique_ptr in scip::create_cons_basic_linear).
+ */
+auto add_tightening_cons(
+	SCIP* scip,
+	xt::xtensor<SCIP_VAR*, 2> const& serving_vars,
+	xt::xtensor<SCIP_VAR*, 1> const& facility_vars) -> void {
+	auto const inf = SCIPinfinity(scip);
+	auto const [n_customers, n_facilities] = serving_vars.shape();
+	assert(facility_vars.size() == n_facilities);
+
```

Comment:
The main tightening constraint, regarding total capacity >= total demand, is missing:
https://github.com/ds4dm/learn2branch/blob/master/01_generate_instances.py#L503

Reply:
You're right